SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp

```diff
@@ -227,19 +246,27 @@ void CDiscAdjMultizoneDriver::Run() {
 
       /*--- Start inner iterations from where we stopped in previous outer iteration. ---*/
 
-      Set_Solution_To_BGSSolution(iZone);
+      Set_Solution_To_BGSSolution_k(iZone);
 
       /*--- Inner loop to allow for multiple adjoint updates with respect to solvers in iZone. ---*/
 
       bool eval_transfer = false;
+      bool no_restart = (iOuterIter > 0) || !config_container[iZone]->GetRestart();
 
-      for (unsigned short iInnerIter = 0; iInnerIter < nInnerIter[iZone]; iInnerIter++) {
+      for (unsigned long iInnerIter = 0; iInnerIter < nInnerIter[iZone]; iInnerIter++) {
 
         config_container[iZone]->SetInnerIter(iInnerIter);
 
         /*--- Add off-diagonal contribution (including the OF gradient) to Solution. ---*/
 
-        Add_External_To_Solution(iZone);
+        if (no_restart || (iInnerIter > 0)) {
+          Add_External_To_Solution(iZone);
+        }
+        else {
+          /*--- If we restarted, Solution already has all contribution,
+           *    we run only one inner iter to compute the cross terms. ---*/
+          eval_transfer = true;
+        }
```

Comment:
On those last commits I tried to enable restarts, it is a bit tricky because we keep solution and cross terms separate during the solution and only combine them when needed. I think the logic is correct, the first outer iteration is used just to extract cross-terms, alas, it does not work perfectly (better than nothing, just not perfect...)